pcapfix.1
.TH PCAPFIX 1 "03 JUL 2021"

.SH "NAME"
pcapfix \- repair pcap and pcapng files

.SH "SYNOPSIS"
\fBpcapfix\fP [\-d] [\-n] [\-o filename] [\-t DATA_LINK_TYPE] [\-v] filename

.SH "DESCRIPTION"
Pcapfix is a tool to repair your damaged or corrupted pcap and pcapng files. It is written in C and released under the GNU General Public License.

To fix your pcap files the tool first checks for an intact pcap global header and repairs it if there are some corrupted bytes. It there seems to be no global header at all, pcapfix adds a self-created one at the beginning of the file. In a second step the tool tries to find pcap packet headers inside the file, below the global header. It checks if the values are correct (or seem to be correct) and tries to repair a packet if there is something wrong.

To fix your pcapng files the tool loops through all packet headers that can be found in the file. It checks for mandatory Section Header and Interface Description Block and creates them if missing. Pcapfix checks for correct block sizes and valid option fields. If something is wrong, invalid fields are repaired (if possible) or skipped and adjusted to finally get a proper pcapng file.

.SH "OPTIONS"
.TP 5
\-d, \-\-deep\-scan
Force deep scan (default = 0)
.br
\-s, \-\-soft\-mode
Switch to soft mode packet detection (default = 0)
.br
This option will force pcapfix to scan for any packet inside the whole file (instead of the first 262144 (max snaplen) bytes only).
.TP
\-n, \-\-pcapng
Force File Format to PCAPNG (default = 0)
.br
This option will force pcapfix to repair the input file as it were in pcapng format (useful for destroyed file headers).
.TP
\-o, \-\-outfile <filename>
Set the output file name. Default will be input file prepended by "fixed_".
.TP
\-t, \-\-data\-link\-type <nr>
Data Link Type (default = 1)
.br
(See NOTES section below.)
.TP
\-v, \-\-verbose
Enable verbose output (default = 0)
.br
You can use multiple \-v options to increase verbosity.
.br
An verbosity of 2 will result in very much output data during package search.

.SH "EXAMPLES"
.TP 5
Repair the pcap file using deep scan mode, because large blocks of the file are corrupted.
pcapfix \-d file.pcap
.TP
Repair the file damaged_file.pcapng using verbose output.
pcapfix \-v damaged_file.pcapng
.TP
Repair the file wlan_traffic.pcap and force the data link type to be 119 (PRISM HEADER).
pcapfix \-t 119 wlan_traffic.pcap
.TP
Repair the file damaged.pcapng which file header is corrupted and force its format to be pcapng.
pcapfix \-n damaged.pcapng

.SH "NOTES"
.TP 5
\fIOutput file name\fP

If you want to perform the reparation process on the input file without creating a separate output file - e.g. in case you are out of space, you can use the -o option and give the same name as the input file.

WARNING: This process is DANGEROUS and may destroy your data. You should only use this option if you know what you are doing and accept the risk of losing the input file.

.TP
\fIDeep scan\fP

In classic pcap files, pcapfix will only scan the first 262144 bytes (maximum packet length) for a proper first packet. If you want to force packet detection even above this limit (e.g. because your file has been heavily destroyed) you can use the deep scan option (\-d).

This option is not necessary for pcapng files because the whole file is arranged in blocks that are 'unlimited' by default. In result pcapfix will always scan the whole pcapng file for further blocks.

.TP
\fISoft mode\fP

There are programs that breach some pcap conventions when writing files. Ignoring those rules do not break the pcap file but they make it difficult for pcapfix to identify the packets in case of broken files.

By default pcapfix uses hard mode. If you encounter problems when repairing files, try to switch to soft mode (-s). Pcapfix will then try to find the packets in a more tolerant way.

.TP
\fIPCAPNG Format\fP

Pcapfix will try to identify the file format to repair (pcap / pcapng) before doing any further checks. If the header itself is corrupted, it will assume the format to be classic pcap. To change this behaviour you can force the tool to do a pcapng-repair by supplying \-n (\-\-pcapng) option.

.TP
\fIVerbosity\fP

You can use multiple \-v options to increase verbosity. An verbosity of 2 will result in very much output data during package search.

.TP
\fIASCII-mode transferred Files (FTP)\fP

Pcapfix is able to repair pcap files that have been transferred in ascii-mode via FTP. In those files a proper pcap structure will be created only to make them readable by wireshark etc. The data inside the packets (and some pcap headers) might still be corrupted. To repair those packets a deeper look inside the packet structure (e.g. checksum) will be necessary.

.TP
\fIData Link Types\fP

You can make pcapfix change / select your data link type by supplying \-t option. Although you may select a data link type number between 0 and 255, only the following types are assigned: If the data link type field is corrupt, pcapfix will select LINKTYPE_ETHERNET by default.

See http://www.tcpdump.org/linktypes.html for further information.

NUMBER	LINK_TYPE

0	LINKTYPE_NULL
.br
1	LINKTYPE_ETHERNET
.br
6	LINKTYPE_TOKEN_RING
.br
7	LINKTYPE_ARCNET_BSD
.br
8	LINKTYPE_SLIP
.br
9	LINKTYPE_PPP
.br
10	LINKTYPE_FDDI
.br
50	LINKTYPE_PPP_HDLC
.br
51	LINKTYPE_PPP_ETHER
.br
100	LINKTYPE_ATM_RFC1483
.br
101	LINKTYPE_RAW
.br
104	LINKTYPE_C_HDLC
.br
105	LINKTYPE_IEEE802_11
.br
107	LINKTYPE_FRELAY
.br
108	LINKTYPE_LOOP
.br
113	LINKTYPE_LINUX_SLL
.br
114	LINKTYPE_LTALK
.br
117	LINKTYPE_PFLOG
.br
119	LINKTYPE_PRISM_HEADER
.br
122	LINKTYPE_IP_OVER_FC
.br
123	LINKTYPE_SUNATM
.br
127	LINKTYPE_IEEE802_11_RADIO
.br
129	LINKTYPE_ARCNET_LINUX
.br
138	LINKTYPE_APPLE_IP_OVER_IEEE1394
.br
139	LINKTYPE_MTP2_WITH_PHDR
.br
140	LINKTYPE_MTP2
.br
141	LINKTYPE_MTP3
.br
142	LINKTYPE_SCCP
.br
143	LINKTYPE_DOCSIS
.br
144	LINKTYPE_LINUX_IRDA
.br
147-162	LINKTYPE_USER0-LINKTYPE-USER15
.br
163	LINKTYPE_IEEE802_11_RADIO_AVS
.br
166	LINKTYPE_PPP_PPPD
.br
169	LINKTYPE_GPRS_LLC
.br
177	LINKTYPE_LINUX_LAPD
.br
187	LINKTYPE_BLUETOOTH_HCI_H4
.br
189	LINKTYPE_USB_LINUX
.br
192	LINKTYPE_PPI
.br
195	LINKTYPE_IEEE802_15_4
.br
196	LINKTYPE_SITA
.br
197	LINKTYPE_ERF
.br
201	LINKTYPE_BLUETOOTH_HCI_H4_WITH_PHDR
.br
202	LINKTYPE_AX25_KISS
.br
203	LINKTYPE_LAPD
.br
204	LINKTYPE_PPP_WITH_DIR
.br
205	LINKTYPE_C_HDLC_WITH_DIR
.br
206	LINKTYPE_FRELAY_WITH_DIR
.br
209	LINKTYPE_IPMB_LINUX
.br
215	LINKTYPE_IEEE802_15_4_NONASK_PHY
.br
220	LINKTYPE_USB_LINUX_MMAPPED
.br
224	LINKTYPE_FC_2
.br
225	LINKTYPE_FC_2_WITH_FRAME_DELIMS
.br
226	LINKTYPE_IPNET
.br
227	LINKTYPE_CAN_SOCKETCAN
.br
228	LINKTYPE_IPV4
.br
229	LINKTYPE_IPV6
.br
230	LINKTYPE_IEEE802_15_4_NOFCS
.br
231	LINKTYPE_DBUS
.br
235	LINKTYPE_DVB_CI
.br
236	LINKTYPE_MUX27010
.br
237	LINKTYPE_STANAG_5066_D_PDU
.br
239	LINKTYPE_NFLOG
.br
240	LINKTYPE_NETANALYZER
.br
241	LINKTYPE_NETANALYZER_TRANSPARENT
.br
242	LINKTYPE_IPOIB
.br
243	LINKTYPE_MPEG_2_TS
.br
244	LINKTYPE_NG40
.br
245	LINKTYPE_NFC_LLCP

.SH "DEVELOPMENT"
This tool is still under development! Please send any further wishes, feature requests or problems in compiling and execution to ruport@f00l.de. Additionally You may send me pcap/pcapng files that could not be repaired too in order to improve pcapfix and get your file repaired.

For further information visit the pcapfix homepage at http://f00l.de/pcapfix/.

.SH "MESSAGES AND EXIT CODES"
 1	, file was corrupted and has been repaired
.br
 0	, file is proper; nothing to repair
.br
\-1	, invalid options / parameters given
.br
\-2	, cannot open input file for reading
.br
\-3	, cannot open output file for writing
.br
\-4	, input file is empty
.br
\-5	, input file is too small
.br
\-6	, file type not supported
.br
\-11	, not a pcap/pcapng file
.br
\-12	, unable to repair the file
.br
\-13	, EOF while reading input file
.br
\-255	, unknown error

.SH "HISTORY"
.TP 5
1.1.7 - 03.07.2021
* fixed security issues
.TP
1.1.6 - 13.06.2021
* added security boundary checks in pcapng format
.TP
1.1.5 - 05.04.2021
* snaplen checks are only performed in soft mode now
.br
* fixed repairs for dlt 195 (ieee 802.15.4 with fcs)
.br
* fixed (binary) file repair bug under Microsoft Windows
.br
* improved compiling on all platforms (thanks to mdeweerd)
.TP
1.1.4 - 22.01.2019
* added support for extended pcap file format (kuznetzov)
.br
* fixed dtl 113 (linux cooked) handling (thanks to Andre Luyer)
.br
* improved hard and soft mode checks for younger/older packets
.br
* minor fixes and improvements
.TP
1.1.3 - 04.09.2018
* implemtented soft mode (for plausibility checks)
.TP
1.1.2 - 16.07.2018
* fixed cross compiling (thanks to Helmut Grohne)
.br
* increased maximum snap length to 262144
.TP
1.1.1 - 25.12.2017
* added write buffers and support to perform repair process without creation of a separate output file
.br
* fixed another MacOS build problem
.br
* fixed minor bugs
.TP
1.1.0 - 31.08.2013
* added checks for valid pcapng format (epb)
.br
* added \-\-outfile parameter to chose fixed file name
.br
* improved pcapng packet alignment (pb, spb, nrb)
.br
* improved pcapng option fields handling
.br
* improved status and verbosity outputs
.br
* fixed reparation bugs with swapped pcap files
.br
* fixed MacOS compile problem
.br
* fixed windows output file name extension missing
.br
* fixed many minor bugs
.TP
1.0.2 - 18.02.2013
* added support for files larger than 2GB on 32bit systems
.TP
1.0.1 - 03.11.2013
* added reparation block type id zero (pcapng)
.br
* added reparation of capture length inside EPB (pcapng)
.br
* set data link type to ethernet on missing header (pcap)
.br
* changed missing pcap header threshold             
.br
* fixed minor bugs
.TP
1.0.0 - 12.10.2013
* added pcapng support
.br
* added nanoseconds support (Issue #1)
.br
* improved console output
.br
* minor bugs fixed
.TP
0.7.3 - 16.06.2013
* added snoop file detection
.br
* added large file support on 32bit architectures
.br
* improved missing header detection
.br
* fixed compiling errors on hurd and kfreebsd architectures
.br
* fixed minor bugs
.TP
0.7.2 - 30.03.2013
* compiles on Apple systems properly now
.br
* fixed problems installing man-pages (on some systems)
.TP
0.7.1 - 03.01.2013
* REALLY fixed file pointer exception on windows64 systems
.br
* updated man-page
.TP
0.7 - 18.10.2012
* added support for swapped (big endian) pcap files
.br
* compiles on OpenBSD properly now
.br
* fixed file pointer exception on windows64 systems
.br
* fixed detection bug when corrupted packet is larger than 65536 bytes
.br
* fixed minimal packet limit to cope with wlan traffic
.TP
0.6 - 20.05.2012
* added deep scan option (\-d) to force packet detection inside the whole file
.br
* detects ascii-corruption in pcap header (unix->win)
.br
* improved global header and packet checks (0 <= usec <= 1000000)
.br
* repair files that first packet is entirely corrupted
.br
* repair oversized packets
.br
* improved last packet mismatch correction
.br
* fixed reading packets over EOF
.TP
0.5 - 05.05.2012
* repair files that packets were not saved chronologically
.br
* detect and repair overlapping packets
.br
* detect and repair cut-off pcap files
.br
* detect and repair ascii-mode transferred pcap files (pcap headers only!)
.br
* added progress bar
.br
* added man-page
.TP
0.4 - 27.04.2012
* completely redesigned packet detection algorithm (replaced bottom-up-recovery with brute-force-packet-guessing)
.br
* improved detection rate by additional plausibility checks
.br
* increased speed when repairing large pcap files
.TP
0.3 - 31.03.2012
* when recovering packets size will be checked to be smaller than 65536
.br
* added recognition when a file does not seem to be a pcap file at all
.br
* compiles on windows systems properly now (tested with dev-cpp)
.br
* added option to manually select data link type
.TP
0.2 - 11.03.2012
* pcapfix compiles on 64bit systems correctly now
.br
* fixed segfault when no filename was given
.br
* fixed (input) file not found bug on directory differ
.br
* added recognition of other data link types beside ethernet in global header
.br
* added source code documentation
.TP
0.1 - 01.03.2012
* this is the first version, everything has changed thou :-)

.SH "COPYRIGHT"
Copyright (c) 2012-2021 Robert Krause

Pcapfix is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or any later version.

Pcapfix is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
.SH "AUTHOR"
Written by Robert Krause <ruport@f00l.de>.


Changelog
pcapfix Changelog
*****************

1.1.7 - 03.07.2021
-------
* fixed security issues

1.1.6 - 13.06.2021
-------
* added security boundary checks in pcapng format

1.1.5 - 05.04.2021
-------
* snaplen checks are only performed in soft mode now
* fixed repairs for dlt 195 (ieee 802.15.4 with fcs)
* fixed (binary) file repair bug under Microsoft Windows
* improved compiling on all platforms (thanks to mdeweerd)

1.1.4 - 22.01.2019
------------------
* added support for extended pcap file format (kuznetzov)
* fixed dtl 113 (linux cooked) handling (thanks to Andre Luyer)
* improved hard and soft mode checks for younger/older packets
* minor fixes and improvements

1.1.3 - 04.09.2018
------------------
* implemtented soft mode (for plausibility checks)

1.1.2 - 16.07.2018
------------------
* fixed cross compiling (thanks to Helmut Grohne)
* increased maximum snap length to 262144

1.1.1 - 25.12.2017
------------------
* added write buffers and support to perform repair
  process without creation of a separate output file
* fixed another MacOS build problem
* fixed minor bugs

1.1.0 - 31.08.2014
------------------
* added checks for valid pcapng format (epb)
* added --outfile parameter to chose fixed file name
* improved pcapng packet alignment (pb, spb, nrb)
* improved pcapng option fields handling
* improved status and verbosity outputs
* fixed reparation bugs with swapped pcap files
* fixed MacOS compile problem
* fixed windows output file name extension missing
* fixed many minor bugs

1.0.2 - 18.02.2014
------------------
* added support for files larger than 2GB on 32bit systems

1.0.1 - 03.11.2013
------------------
* added reparation block type id zero (pcapng)
* added reparation of capture length inside EPB (pcapng)
* set data link type to ethernet on missing header (pcap)
* changed missing pcap header threshold
* fixed minor bugs

1.0.0 - 12.10.2013
------------------
* added pcapng support
* added nanoseconds support (Issue #1)
* improved console output
* minor bugs fixed

0.7.3 - 16.06.2013
------------------
* added snoop file detection
* added large file support on 32bit architectures
* improved missing header detection
* fixed compiling errors on hurd and kfreebsd architectures
* fixed minor bugs

0.7.2 - 30.03.2013
------------------
* compiles on Apple systems properly now
* fixed problems installing man-pages (on some systems)

0.7.1 - 03.01.2013
------------------
* REALLY fixed file pointer exception on windows64 systems
* updated man-page

0.7 - 18.10.2012
----------------
* added support for swapped (big endian) pcap files
* compiles on OpenBSD properly now
* fixed file pointer exception on windows64 systems
* fixed detection bug when corrupted packet is larger than 65536 bytes
* fixed minimal packet limit to cope with wlan traffic

0.6 - 20.05.2012
----------------
* added deep scan option (-d) to force packet detection inside the whole file
* detects ascii-corruption in pcap header (unix->win)
* improved global header and packet checks (0 <= usec <= 1000000)
* repair files that first packet is entirely corrupted
* repair oversized packets
* improved last packet mismatch correction
* fixed reading packets over EOF

0.5 - 05.05.2012
----------------
* repair files that packets were not saved chronologically
* detect and repair overlapping packets
* detect and repair cut-off pcap files
* detect and repair ascii-mode transfered pcap files (pcap headers only!)
* added progress bar
* added man-page

0.4 - 27.04.2012
----------------
* completely redesigned packet detection algorithm (replaced bottom-up-recovery with brute-force-packet-guessing)
* improved detection rate by additional plausability checks
* increased speed when repairing large pcap files

0.3 - 31.03.2012
----------------
* when recovering packets size will be checked to be smaller than 65536
* added recognition when a file does not seem to be a pcap file at all
* compiles on windows systems properly now (tested with dev-cpp)
* added option to manually select data link type

0.2 - 11.03.2012
----------------
* pcapfix compiles on 64bit systems correctly now
* fixed segfault when no filename was given
* fixed (input) file not found bug on directory differ
* added recognition of other data link types beside ethernet in global header
* added source code documentation

0.1 - 01.03.2012
----------------
* this is the first version, everything has changed thou :-)


pcapfix.c
/*******************************************************************************
 *
 * Copyright (c) 2012-2021 Robert Krause (ruport@f00l.de)
 *
 * This file is part of Pcapfix.
 *
 * Pcapfix is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or any later version.
 *
 * Pcapfix is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Pcapfix. If not, see http://www.gnu.org/licenses/.
 *
 * Last Modified: 03.07.2021
 *
 *******************************************************************************
 *
 * Description:
 *
 * Pcapfix is a tool to repair your damaged or corrupted pcap and pcapng files.
 * It is written in C and released under the GNU General Public License.
 *
 * To fix your pcap files the tool first checks for an intact pcap global header
 * and repairs it if there are some corrupted bytes. It there seems to be no
 * global header at all, pcapfix adds a self-created one at the beginning of the
 * file. In a second step the tool tries to find pcap packet headers inside the
 * file, below the global header. It checks if the values are correct (or seem
 * to be correct) and tries to repair a packet if there is something wrong. 
 *
 * To fix your pcapng files the tool loops through all packet headers that can
 * be found in the file. It checks for mandatory Section Header Block and
 * Interface Description Block and creates them if missing. Pcapfix checks for
 * correct block sizes and valid option fields. If something is wrong, invalid
 * fields are repaired (if possible) or skipped and adjusted to finally get a
 * proper pcapng file.
 *
 ******************************************************************************/

#include "pcapfix.h"
#include "pcap.h"
#include "pcap_kuznet.h"
#include "pcapng.h"

#define VERSION "1.1.7"		    /* pcapfix version */

#define BTSNOOP_MAGIC 0x6E737462    /* btsnoop file magic (first 4 bytes) */
#define SNOOP_MAGIC 0x6f6f6e73	    /* snoop file magic (first 4 bytes) */
#define NETMON_MAGIC 0x55424d47     /* netmon file magic */
#define NETMON11_MAGIC 0x53535452   /* netmon 1.1 file magic */
#define ETHERPEEK_MAGIC 0x7265767f  /* EtherPeek/AiroPeek/OmniPeek file magic */

/* configuration variables */
int deep_scan = 0;		/* deep scan option (default: no deep scan) */
int soft_mode = 0;		/* soft mode option (default: no soft mode) */
int keep_outfile = 0;		/* keep output file even if nothing needed fixing (default: don't) */
int verbose = 0;		/* verbose output option (default: dont be verbose) */
int swapped = 0;		/* pcap file is swapped (big endian) */
int data_link_type = -1;	/* data link type (default: LINKTYPE_ETHERNET) */
int pcapng = 0;			/* file format to assume */

/* header placeholder */
unsigned int header_magic;

/*
 * Function:  usage
 * ----------------
 * prints out the usage information of pcapfix
 * invoked when running without / with too less parameters
 *
 * in:  the name of pcapfix' binary file
 *
 */
void usage(char *progname) {
  printf("Usage: %s [OPTIONS] filename\n", progname);
  printf("OPTIONS:");
  printf(  "\t-d        , --deep-scan          \tDeep scan (pcap only)\n");
  printf("\t\t-s        , --soft-mode          \tSoft mode (packet detection)\n");
  printf("\t\t-n        , --pcapng             \tforce pcapng format\n");
  printf("\t\t-o <file> , --outfile <file>     \tset output file name\n");
  printf("\t\t-k        , --keep-outfile       \tdon't delete the output file if nothing needed to be fixed\n");
  printf("\t\t-t <nr>   , --data-link-type <nr>\tData link type\n");
  printf("\t\t-v        , --verbose            \tVerbose output\n");
  printf("\n");
}

/*
 * Function:  conshort
 * -------------------
 * Convertes a short variable into correct byte order (if necessary)
 * depending on the global -swapped- settings of pcap data.
 *
 * This function MUST encapsulate all variables that might be swapped!
 *
 * var: the variable that shall be converted into correct byte order
 *
 * returns: the variable in correct byte order
 *
 */
unsigned short conshort(unsigned short var) {
  if (swapped == 0) return(var);
  return(htons(var));
}

/*
 * Function:  conint
 * -----------------
 * Convertes an integer variable into correct byte order (if necessary)
 * depending on the global -swapped- settings of pcap data.
 *
 * This function MUST encapsulate all variables that might be swapped!
 *
 * var: the variable that shall be converted into correct byte order
 *
 * returns: the variable in correct byte order
 *
 */
unsigned int conint(unsigned int var) {
  if (swapped == 0) return(var);
  return(htonl(var));
}

/*
 * Function:  print_progress
 * -------------------------
 * prints the progess line when using pcapfix in non-verbose mode
 *
 * pos:       the current filepointer position
 * filesize:  the size of the input pcap file in bytes
 *
 */
void print_progress(off_t pos, off_t filesize) {
  float percentage;	/* pencentage variable */

  /* calculate the current percentage of file analyzing progress */
  percentage = (float)(100*(double)pos/(double)filesize);

  /* print the first part of the line including percentage output */
  printf("[*] Progress: %6.02f %%\n", percentage);
}

/*
 * Function:  main
 * ---------------
 * - takes arguments from command line
 * - opens input and output files
 * - checks for file format (pcap, pcapng, snoop, ...)
 * - runs fixing function depending on format
 * - evaluates result
 *
 * argc:  number of cmd line args
 * argv:  array of pointers to cmd line args
 *
 * returns:  1    success (filed repaired)
 *           0    success (nothing to fix)
 *          -1    error (parameters)
 *          -2    error (cannot open input file)
 *          -3    error (cannot open output file for writing)
 *          -4    error (input file is empty)
 *          -5    error (input file is too small)
 *          -6    error (filetype is known but not supported)
 *          -11   error (not a pcap file)
 *          -12   error (unable to recover pcap file)
 *          -13   error (EOF while reading input file)
 *          -255  error (unkown)
 *
 */
int main(int argc, char *argv[]) {
  FILE *pcap, *pcap_fix;		/* input and output file */
  int option_index = 0;			/* getopt_long option index */
  int c;                        /* getopts loop counter */
  int res;                      /* return values */
  char *filename;               /* filename of input file */
  char *filebname;              /* filebasename of input file (without path) */
  char *filename_fix = NULL;    /* filename of output file */
  off_t bytes;		        /* read/written blocks counter */
  off_t filesize;	        /* file size of input pcap file in bytes */

  /* init getopt_long options struct */
  struct option long_options[] = {
    {"deep-scan", no_argument, 0, 'd'},            /* --deep-scan == -d */
    {"soft-mode", no_argument, 0, 's'},            /* --soft-mode == -s */
    {"pcapng", no_argument, 0, 'n'},               /* --pcapng == -n */
    {"outfile", required_argument, 0, 'o'},        /* --outfile == -o */
    {"keep-outfile", no_argument, 0, 'k'},         /* --keep-outfile == -k */
    {"data-link-type", required_argument, 0, 't'}, /* --data-link-type == -t */
    {"verbose", no_argument, 0, 'v'},              /* --verbose == -v */
    {0, 0, 0, 0}
  };

  /* print out pcapfix header information */
  printf("pcapfix %s (c) 2012-2021 Robert Krause\n\n", VERSION);

  /* scan for options and arguments */
  while ((c = getopt_long(argc, argv, ":t:ko:v::d::s::n::", long_options, &option_index)) != -1) {
    switch (c) {
      case 0:	/* getopt_long options evaluation */
        break;
      case 'd':	/* deep scan */
        deep_scan++;
        break;
      case 's':	/* soft mode */
        soft_mode++;
        break;
      case 'k': /* keep outfile even if nothing needed fixing */
        keep_outfile++;
        break;
      case 'n':	/* pcapng format */
        pcapng++;
        break;
      case 'o':	/* output file name */
        filename_fix = malloc(strlen(optarg)+1);
	strcpy(filename_fix, optarg);
        break;
      case 't':	/* data link type */
        data_link_type = atoi(optarg);
        break;
      case 'v':	/* verbose */
        verbose++;
        break;
      case '?': /* unknown option */
        usage(argv[0]);
        return -1;
      default:
        abort();
    }
  }

  /* filename is first argument */
  filename = argv[optind++];

  /* if filename is not set, output usage information */
  if (filename == NULL) {
    if (filename_fix) {
      free(filename_fix);
    }
    usage(argv[0]);
    return(-1);
  }

  /* open input file */
  printf("[*] Reading from file: %s\n", filename);
  pcap = fopen(filename, "rb");
  if (!pcap) {
    perror("[-] Cannot open input file");
    if (filename_fix) {
      free(filename_fix);
    }
    return(-2);
  }

  /* check for preassigned fixed file name */
  if (filename_fix == NULL) {

    /* open output file */
    /* we need to extract the basename first (windows and linux use different functions) */
    filebname = malloc(strlen(filename)+1);
    #ifdef __WIN32__
      char *fileext = malloc(strlen(filename));   /* file extention to be used in output file as well */
      _splitpath(filename, NULL, NULL, filebname, fileext);	/* windown method (_splitpath) */
      strcat(filebname, fileext);
      free(fileext);
    # else
      strcpy(filebname, basename(filename));		/* unix method (basename) */
    #endif
    filename_fix = malloc(strlen(filebname)+7);	/* size of outputfile depends on inputfile's length */

    /* prepare output file name: "fixed_" + inputfilename */
    strcpy(filename_fix, "fixed_");
    strcat(filename_fix, filebname);
    free(filebname);
  }

  /* open the file for writing */

  // is output == inputfile ?? if yes, then open for read / append
  // if no, then open for writing only
  if (strcmp(filename, filename_fix) == 0) pcap_fix = fopen(filename_fix, "rb+");
  else pcap_fix = fopen(filename_fix, "wb+");

  if (!pcap_fix) {
    perror("[-] Cannot open output file for writing");
    return(-3);
  }
  printf("[*] Writing to file: %s\n", filename_fix);

  /* basic checks of input file */

  /* get file size */
  fseeko(pcap, 0, SEEK_END);
  filesize = ftello(pcap);
  printf("[*] File size: %" FMT_OFF_T " bytes.\n", filesize);

  /* check for empty file */
  if (filesize == 0) {
    printf("[-] The source file is empty.\n\n");
    fclose(pcap);
    fclose(pcap_fix);

    if (strcmp(filename, filename_fix) != 0) remove(filename_fix);
    free(filename_fix);

    return(-4);
  }

  /* reset file pointer to file begin */
  fseeko(pcap, 0, SEEK_SET);

  /* read header to header magic for further inspection */
  bytes = fread(&header_magic, sizeof(header_magic), 1, pcap);
  if (bytes == 0) {
    printf("[-] Cannot read file header (file too small?).\n\n");
    fclose(pcap);
    fclose(pcap_fix);

    if (strcmp(filename, filename_fix) != 0) remove(filename_fix);
    free(filename_fix);

    return(-5);
  }
  fseeko(pcap, 0, SEEK_SET);

  /* check for file type */
  switch (header_magic) {

    /* etherpeek file format --> often used with pcapfix but NOT supported (yet) */
    case ETHERPEEK_MAGIC:
      printf("[-] This is a EtherPeek/AiroPeek/OmniPeek file, which is not supported.\n\n");

      /* close input and output files */
      fclose(pcap);
      fclose(pcap_fix);

      /* delete output file due to no changes failure */
      if (strcmp(filename, filename_fix) != 0) remove(filename_fix);
      free(filename_fix);

      return(-6);

    /* netmon file format --> often used with pcapfix but NOT supported (yet) */
    case NETMON_MAGIC:
    case NETMON11_MAGIC:
      printf("[-] This is a NetMon file, which is not supported.\n\n");

      /* close input and output files */
      fclose(pcap);
      fclose(pcap_fix);

      /* delete output file due to no changes failure */
      if (strcmp(filename, filename_fix) != 0) remove(filename_fix);
      free(filename_fix);

      return(-6);

    /* SNOOP file format --> often used with pcapfix but NOT supported (yet) */
    case SNOOP_MAGIC:
      printf("[-] This is a SNOOP file, which is not supported.\n\n");

      /* close input and output files */
      fclose(pcap);
      fclose(pcap_fix);

      /* delete output file due to no changes failure */
      if (strcmp(filename, filename_fix) != 0) remove(filename_fix);
      free(filename_fix);

      return(-6);

    case BTSNOOP_MAGIC:
      printf("[-] This is a BTSNOOP file, which is not supported.\n\n");

      /* close input and output files */
      fclose(pcap);
      fclose(pcap_fix);

      /* delete output file due to no changes failure */
      if (strcmp(filename, filename_fix) != 0) remove(filename_fix);
      free(filename_fix);

      return(-6);

    /* extended pcap format (KUZNETZOV) */
    case PCAP_EXT_MAGIC:
    case PCAP_EXT_MAGIC_SWAPPED:
      printf("[+] This is an extended tcpdump file.\n");
      res = fix_pcap_kuznetzov(pcap, pcap_fix);

      break;

    /* PCAPNG format */
    case PCAPNG_MAGIC:
      printf("[+] This is a PCAPNG file.\n");
      res = fix_pcapng(pcap, pcap_fix);
      break;

    /* classic PCAP format */
    case PCAP_MAGIC:
    case PCAP_MAGIC_SWAPPED:
      printf("[+] This is a PCAP file.\n");
      if (pcapng > 0) {
        printf("[!] Your wish is my command! I will handle it as PCAPNG nevertheless.\n");
        res = fix_pcapng(pcap, pcap_fix);
      } else {
        res = fix_pcap(pcap, pcap_fix);
      }
      break;

    /* if the file type is unknown (maybe header corrupted) assume classic PCAP format */
    default:
      if (pcapng > 0) {
        printf("[*] Unknown file type. Assuming PCAPNG format.\n");
        res = fix_pcapng(pcap, pcap_fix);
      } else {
        printf("[*] Unknown file type. Assuming PCAP format.\n");
        res = fix_pcap(pcap, pcap_fix);
      }
      break;
  }

  /* evaluate result of fixing function */
  switch (res) {

    /* no corruption found; all fields were valid */
    case 0:
      printf("[*] Your pcap file looks proper. Nothing to fix!\n\n");

      /* close input and output files */
      fclose(pcap);
      fclose(pcap_fix);

      /* delete output file due to no changes failure */
      if ((strcmp(filename, filename_fix) != 0) && (0 == keep_outfile)) remove(filename_fix);
      free(filename_fix);

      return(0);

    /* there is NO indication that this has ever been a pcap file at all */
    case -1:
      printf("[-] FAILED: This file does not seem to be a pcap/pcapng file!\n\n");

      /* close input and output files */
      fclose(pcap);
      fclose(pcap_fix);

      /* delete output file due to failure */
      if (strcmp(filename, filename_fix) != 0) remove(filename_fix);
      free(filename_fix);

      /* deep scan / soft mode dependent output */
      if (deep_scan == 0 || soft_mode == 0) printf("If you are SURE that there are pcap packets inside, try with deep scan (-d) (pcap only!) and/or soft mode (-s) to find them!\n\n");
      else printf("There is REALLY no pcap packet inside this file!!!\n\n");

      return(res-10);

    /* it seems to be a pcap file, but pcapfix can NOT repair it (yet) */
    case -2:
      printf("[-] FAILED: Unable to recover pcap file.\n\n");

      /* some hints for verbose mode and pcapfix improvement support */
      if (!verbose) printf("Try executing pcapfix with -v option to trace the corruption!\n");
      printf("You may help to improve pcapfix by sending your pcap file to ruport@f00l.de\n\n");

      /* close input and output files */
      fclose(pcap);
      fclose(pcap_fix);

      /* delete output file due to failure */
      if (strcmp(filename, filename_fix) != 0) remove(filename_fix);
      free(filename_fix);

      return(res-10);

    /* fread did not succeed; might be caused by early EOF; but it should NOT happen at all */
    case -3:
      printf("[-] FAILED: EOF while reading input file.\n\n");

      /* some hints for verbose mode and pcapfix improvement support */
      if (!verbose) printf("Try executing pcapfix with -v option to trace the corruption!\n");
      printf("You may help to improve pcapfix by sending your pcap file to ruport@f00l.de\n\n");

      /* close input and output files */
      fclose(pcap);
      fclose(pcap_fix);

      /* delete output file due to failure */
      if (strcmp(filename, filename_fix) != 0) remove(filename_fix);
      free(filename_fix);

      return(res-10);
  }

  /* file has been progressed properly. what is the result (number of corruptions)? */

  if (res > 0) {
    /* Successful repaired! (res > 0) */

    fclose(pcap);
    off_t finalpos = ftello(pcap_fix);
    fclose(pcap_fix);
    int success = truncate(filename_fix, finalpos);
    if (success != 0) printf("[-] Truncating result file failed!");
    free(filename_fix);

    printf("[+] SUCCESS: %d Corruption(s) fixed!\n\n", res);
    return(0);

  } else {
    /* Unknown Error (res < 0); this should NEVER happen! */

    printf("[-] UNKNOWN ERROR (%d)\n\n", res);
    printf("Please report this bug to ruport@f00l.de (including -v -v output).\n\n");
    return(-255);

  }

}
